pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go

```diff
@@ -0,0 +1,523 @@
+package aggregators
+
+import (
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/shopspring/decimal"
+	"google.golang.org/protobuf/proto"
+
+	ocrcommon "github.com/smartcontractkit/libocr/commontypes"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities/consensus/ocr3/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+	"github.com/smartcontractkit/chainlink-common/pkg/values/pb"
+)
+
+const (
+	AGGREGATION_METHOD_MEDIAN = "median"
+	AGGREGATION_METHOD_MODE   = "mode"
+	DEVIATION_TYPE_NONE       = "none"
+	DEVIATION_TYPE_PERCENT    = "percent"
+	DEVIATION_TYPE_ABSOLUTE   = "absolute"
+	REPORT_FORMAT_MAP         = "map"
+	REPORT_FORMAT_ARRAY       = "array"
+	REPORT_FORMAT_VALUE       = "value"
+
+	DEFAULT_REPORT_FORMAT     = REPORT_FORMAT_MAP
+	DEFAULT_OUTPUT_FIELD_NAME = "Reports"
+)
+
+type ReduceAggConfig struct {
+	// Configuration on how to aggregate one or more data points
+	Fields []AggregationField `mapstructure:"fields"  required:"true"`
+	// The top level field name that report data is put into
+	OutputFieldName string `mapstructure:"outputFieldName" json:"outputFieldName" default:"Reports"`
+	// The structure surrounding the report data that is put on to "OutputFieldName"
+	ReportFormat string `mapstructure:"reportFormat" json:"reportFormat" default:"map" jsonschema:"enum=map,enum=array,enum=value"`
+	// Optional key name, that when given will contain a nested map with designated Fields moved into it
+	// If given, one or more fields must be given SubMapField: true
+	SubMapKey string `mapstructure:"subMapKey" json:"subMapKey" default:""`
+}
+
+type AggregationField struct {
+	// An optional check to only report when the difference from the previous report exceeds a certain threshold.
+	// Can only be used when the field is of a numeric type: string, decimal, int64, big.Int, time.Time, float64
+	// If no deviation is provided on any field, there will always be a report once minimum observations are reached.
+	Deviation       decimal.Decimal `mapstructure:"-"  json:"-"`
+	DeviationString string          `mapstructure:"deviation"  json:"deviation,omitempty"`
+	// The format of the deviation being provided
+	// * percent - a percentage deviation
+	// * absolute - an unsigned numeric difference
+	DeviationType string `mapstructure:"deviationType" json:"deviationType,omitempty" jsonschema:"enum=percent,enum=absolute,enum=none"`
+	// The key to find a data point within the input data
+	// If omitted, the entire input will be used
+	InputKey string `mapstructure:"inputKey" json:"inputKey"`
```

Comment:
is this a key or a path? ie for nested structures you will get map of maps, maybe Path would be a better name?